compression/provider.go

```diff
@@ -0,0 +1,212 @@
+package compression
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+
+	"github.com/DataDog/zstd"
+	lz4 "github.com/cloudflare/golz4"
+	"github.com/klauspost/compress/s2"
+)
+
+// NoCompressionService uses no compression
+var NoCompressionService *noCompression = &noCompression{
+	id: 0,
+}
+
+// ZstdCompressionService uses  zstd compression method
+// github.com/DataDog/zstd
+var ZstdCompressionService *zstdCompression = &zstdCompression{
+	id:               1,
+	compressionLevel: 3,
+}
+
+// S2CompressionService uses s2/snappy compression method
+// github.com/klauspost/compress/s2
+var S2CompressionService *s2Compression = &s2Compression{
+	id: 2,
+}
+
+// Lz4CompressionService uses lz4 compression
+// github.com/cloudflare/golz4
+var Lz4CompressionService *lz4Compression = &lz4Compression{
+	id: 3,
+}
+
+type noCompression struct {
+	id byte
+}
+
+// Compress returns src without any changes.
+func (c noCompression) Compress(src []byte) ([]byte, error) {
+	return append(src, c.id), nil
+}
+
+// Decompress returns src without any changes.
+func (c noCompression) Decompress(src []byte) ([]byte, error) {
+	return src[:len(src)-providerIDLengthInByte], nil
+}
+
+// GetID returns compression identifier.
+func (c noCompression) GetID() byte {
+	return c.id
+}
+
+type zstdCompression struct {
+	id               byte
+	compressionLevel int
+}
+
+// Compress compresses src  using zstd method
+func (c zstdCompression) Compress(src []byte) ([]byte, error) {
+
+	output, err := zstd.CompressLevel(nil, src, c.compressionLevel)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata, err := getMetadata(c.id, len(src))
```

Comment:
I think that metadata should be added by the engine. It will simplify the providers itself